### PR TITLE
[AnimeLab] Add new extractor

### PR DESCRIPTION
**Pull request by mariuszskon**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This pull request adds basic support for the Australian / New Zealand streaming site [AnimeLab](https://www.animelab.com). More features are planned, but I wanted to get a basic version out for now.

The site is run by [Madman Entertainment](https://www.animelab.com/terms-conditions), and is legal.

The site requires authentication, but a free account is simple to make (only requres an email address, which it does not even need a confirmation actually exists) and can access most of the content (but the site is geo-restricted to Aus/NZ).

It would be great if someone has a premium account and can verify that premium content is accessible (and does not interfere with the tests).

Plase leave feedback on what needs to be changed :)

### CODE REVIEW

File: youtube_dl/extractor/animelab.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            response = self._download_webpage(
+                self._LOGIN_URL, None, 'Logging in', 'Wrong login info',
+                data=urlencode_postdata(login_form),
+                headers={'Content-Type': 'application/x-www-form-urlencoded'})
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 400:
+                page = e.cause.read().decode('utf-8')
+                js_error = self._search_regex(
+                    r'Utils\.notify\( *?[\'"]error[\'"] *?, *?[\'"](.*?)[\'"]',
+                    page, 'Trying to get error message in js', default=None, fatal=False)
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: `default` is not used with `fatal`.

File: youtube_dl/extractor/animelab.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        if is_logged_in(response):
+            return
+
+        raise ExtractorError('Unable to login (cannot verify if logged in)')
+
+    def _real_initialize(self):
+        self._login()
+
+    def _extract_video_collection_from_player(self, webpage):
+        video_collection_str = self._search_regex(r'new +?VideoCollection *?\((.*?)\);', webpage, 'AnimeLab VideoCollection')
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: `\s` not ` `.

File: youtube_dl/extractor/animelab.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        return self._parse_json(video_collection_str, None)
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: `None` is not a `video_id`.

File: youtube_dl/extractor/animelab.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+    def _extract_position_from_player(self, webpage):
+        position_str = self._search_regex(r'playlistPosition *?= *?(\d+)', webpage, 'Playlist Position')
+        position = int_or_none(position_str)
+        if position is None:
+            raise ExtractorError('Could not get playlist position as integer. Extracted string was %s' % position_str)
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: This is checked by `_search_regex`.

File: youtube_dl/extractor/animelab.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+        return position
+
+    def _create_title(self, webpage, title_parts):
+        if None not in title_parts:
+            return '%s - %s %s - %s' % title_parts
+        else:
+            self.report_warning('Could not get all necessary data to contruct title manually, trying alternative sources...')
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: Remove.

File: youtube_dl/extractor/animelab.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                format_id = '_'.join([x for x in format_id_parts if x is not None])
+                current_format['format_id'] = format_id
+
+                formats.append(current_format)
+
+        # due to Python's sorting algorithm having the property of being 'stable', the following
+        # means that videos are sorted by height, but in the case of the same height,
+        # are sorted by language
+        formats.sort(key=lambda k: str.lower(k.get('language') or ''))
+        formats.sort(key=lambda k: k.get('height') or 0)
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: `_sort_formats`.

File: youtube_dl/extractor/animelab.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            'title': title,
+            'description': description,
+            'series': series,
+            'episode': episode_name,
+            'episode_number': int_or_none(episode_number),
+            'formats': formats,
+        }
+
+    def _extract_video_from_player(self, webpage):
+        video_collection = self._extract_video_collection_from_player(webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: Eliminate all methods that is only used once.

File: youtube_dl/extractor/animelab.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        position = self._extract_position_from_player(webpage)
+        return self._extract_video_from_collection(webpage, video_collection, position)
+
+
+class AnimeLabIE(AnimeLabBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?animelab\.com/player/.+'
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: No `id` extracted.

File: youtube_dl/extractor/animelab.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            'formats': formats,
+        }
+
+    def _extract_video_from_player(self, webpage):
+        video_collection = self._extract_video_collection_from_player(webpage)
+        position = self._extract_position_from_player(webpage)
+        return self._extract_video_from_collection(webpage, video_collection, position)
+
+
+class AnimeLabIE(AnimeLabBaseIE):
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: No point in base class.

File: youtube_dl/extractor/animelab.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    _VALID_URL = r'https?://(?:www\.)?animelab\.com/player/.+'
+
+    # the following tests require authentication, but a free account will suffice
+    # just set 'netrc' to true in test/local_parameters.json if you use a .netrc file
```
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: Skip tests.

File: youtube_dl/extractor/animelab.py (hunk 8 above)
Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: That is because there is no numeric id in the URL (there is a slug title, is this what I should use as id? (see tests for example))

File: youtube_dl/extractor/animelab.py (hunk 5 above)
Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: So I should rely on title from HTML alone (I'm just not sure exactly what to remove)?

File: youtube_dl/extractor/animelab.py (hunk 9 above)
Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: You're right, in this case the base class is quite pointless, but if I want to add playlist support in the future, a base class just for login will be useful (so I will make the base class smaller rather than completely removing it).

File: youtube_dl/extractor/animelab.py (hunk 3 above)
Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: Will fix once I can get id from url (see other question about `id`)

**Comment by mariuszskon** [comment]
I have resolved the issues I could so far. When you reply to my other queries, I will solve the related issues. Hopefully I made the code easier to read by inlining the methods I only used once :)

File: youtube_dl/extractor/animelab.py (hunk 3 above)
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: Use display id.

File: youtube_dl/extractor/animelab.py (hunk 5 above)
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: You are providing a fallback for title extraction thus you **assume** this situation normal and should not print any warning.

File: youtube_dl/extractor/animelab.py (hunk 9 above)
Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: You will add it when there will be a playlist support. For now it's completely useless.

File: youtube_dl/extractor/animelab.py (hunk 8 above)
Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: @dstftw You did not reply to this question - how should I handle this situation (is getting it as `display_id` enough, and then getting the `id` through the json)?

Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: It appears that the [khanacademy](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/extractor/khanacademy.py#L19) extractor uses a slug title as its `id`. Should I do the same?

Comment (dstftw) [inline youtube_dl/extractor/animelab.py]: I've already pointed out: use `display_id` until you get real id.

Comment (mariuszskon) [inline youtube_dl/extractor/animelab.py]: OK, sorry.

**Comment by mariuszskon** [comment]
@dstftw I believe I have resolved the issues. Please review at your convenience.

**Comment by mariuszskon** [comment]
@dstftw Bump. I also removed a superflous test and added thumbnail extraction. In case I was a bit unclear, I plan on adding playlist extraction once this is merged (that is, I am ready for this to be merged when you have reviewed my submission). Thanks for your review in advance.

**Comment by mariuszskon** [comment]
@dstftw Could I ask for a re-review? Quite a bit has changed ;)
Thanks.

**Comment by mariuszskon** [comment]
Since it seems like this will not get merged any time soon, I decided to add some features, most notably, playlist support. Please review at your convenience.

**Comment by XxUnkn0wnxX** [comment]
is support to animelab ever going to come? funimation is going to close down here in AUS soon, & i'd rather not be bothered with a VPN when animelab could be a good alternative source to DL from.

**Comment by mariuszskon** [comment]
@XxUnkn0wnxX 

Quite a few things have changed since 2017.

Luckily I decided to update everything :)

Of course no guarantee this will be merged into official youtube-dl
In the mean time, feel free to use my fork https://github.com/mariuszskon/youtube-dl/tree/animelab

**Comment by inks007** [comment]
@mariuszskon Thanks for writing this extractor.

I'm doing some testing with mariuszskon's branch.
The downloader works well. I've noted 2 issues:

Downloading playlists triggers an error if you're using --download-archive
```
[download] Downloading playlist: My Hero Academia
[AnimeLabShows] playlist My Hero Academia: Collected 89 video ids (downloading 89 of them)
[download] Downloading video 1 of 89
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/xxxx/youtube-dl/youtube_dl/__main__.py", line 19, in <module>
    youtube_dl.main()
  File "/home/xxxx/youtube-dl/youtube_dl/__init__.py", line 474, in main
    _real_main(argv)
  File "/home/xxxx/youtube-dl/youtube_dl/__init__.py", line 464, in _real_main
    retcode = ydl.download(all_urls)
  File "/home/xxxx/youtube-dl/youtube_dl/YoutubeDL.py", line 2018, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "/home/xxxx/youtube-dl/youtube_dl/YoutubeDL.py", line 807, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/home/xxxx/youtube-dl/youtube_dl/YoutubeDL.py", line 1000, in process_ie_result
    reason = self._match_entry(entry, incomplete=True)
  File "/home/xxxx/youtube-dl/youtube_dl/YoutubeDL.py", line 752, in _match_entry
    if self.in_download_archive(info_dict):
  File "/home/xxxx/youtube-dl/youtube_dl/YoutubeDL.py", line 2100, in in_download_archive
    vid_id = self._make_archive_id(info_dict)
  File "/home/xxxx/youtube-dl/youtube_dl/YoutubeDL.py", line 2093, in _make_archive_id
    return extractor.lower() + ' ' + video_id
TypeError: must be str, not int
```

The second are the following warnings:
```
WARNING: [AnimeLab] Could not get height of video
```
These don't seem to impact the file format selectors, e.g. -f '[height>720]'

It would be great if youtube-dl would merge this extractor. Animelab's new web player is buggy.

**Comment by BelSean21** [comment]
Hi @mariuszskon.

I'm not familiar with Python or youtube-dland I'm trying to get this to work on an RPi, Make seems to have successfully created the executable inspite of some errors relating to readme files. However, when I attempt to use it I'm getting an unsupported URL errors

```
pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $ youtube-dl -u xxxxxx@xxxxxx.xxx.au https://www.animelab.com/player/cross-ange-rondo-of-angels-and-dragons-episode-1
Type account password and press [Return]: 
[generic] cross-ange-rondo-of-angels-and-dragons-episode-1: Requesting header
WARNING: Falling back on generic information extractor.
[generic] cross-ange-rondo-of-angels-and-dragons-episode-1: Downloading webpage
[generic] cross-ange-rondo-of-angels-and-dragons-episode-1: Extracting information
ERROR: Unsupported URL: https://www.animelab.com/player/cross-ange-rondo-of-angels-and-dragons-episode-1
pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $
```
`Clearly I'm missing something, I'd be grateful for any light you could shed

Thanks, Sean.

**Comment by inks007** [comment]
@BelSean21 Did you compile mariuszskon's version? With the correct animelab branch?

youtube-dl --list-extractors
If the animelab extractor is missing, you have the wrong branch or wrong build.

youtube-dl --version
The version might give you clues as to which build you are invoking.

**Comment by BelSean21** [comment]
Hi @inks007,

`pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $ ./youtube-dl --list-extractors > extract.lst
pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $ ./youtube-dl --version
2020.01.24
`
I don't see Animelab in the extractor list, which suggests I have the wrong build.

I cloned mariuszskon's Git repository to a local directory and ran make and got this binary. Could make be picking up the standard version that's also installed?

Thanks,
Sean.

**Comment by inks007** [comment]
@BelSean21 I made the same mistake.
git checkout -b animelab

**Comment by BelSean21** [comment]
@inks007 I'm still getting the same result. Make still appears to be compiling the Master branch

**Comment by inks007** [comment]
I didn't bother building it.
Try "python3 -m youtube_dl" from the cloned directory. If youtube_dl/extractor/animelab.py exists, then you've got the correct branch.

**Comment by BelSean21** [comment]
@inks007 
Turns out youtube_dl/extractor/animelab.py was missing and I had the wrong branch.

`git clone -b animelab https://github.com/mariuszskon/youtube-dl.git`

Got me the correct one. I now have it working. 

Thanks for your help.
Sean.

**Comment by mariuszskon** [comment]
Hey @inks007 and @BelSean21, glad you guys are enjoying my extractor! Fingers crossed it eventually gets merged into master for youtube-dl, but of course no guarantees.

Regarding your issues @inks007, I'll take a closer look when I get the time. A quick glance at the stack trace for `--download-archive` (your error message) shows that the AnimeLab extractor code is never involved so I am not sure if the issue is with the extractor...
The height warning should be an easy fix on my end once I have a look at the website again.

**Comment by BelSean21** [comment]
@mariuszskon Thanks for creating it

**Comment by mariuszskon** [comment]
@inks007 the problem with `--download-archive` has been fixed in the latest commit 395f0e7. Please update and tell me know if all is good on your end too :)

**Comment by mariuszskon** [comment]
@inks007 I cannot reproduce your issue regarding getting heights:
```
$ python -m youtube_dl --netrc https://www.animelab.com/player/my-hero-academia-episode-1
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] my-hero-academia-episode-1: Downloading requested URL
[download] Destination: My Hero Academia - Episode 1 - Izuku Midoriya - Origin-6401.mp4
```
If you could provide an exact test command/URL that causes this behaviour that would be awesome.

**Comment by BelSean21** [comment]
Hey @mariuszskon I'm seeing the same issue as @inks007

```
pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $ ./youtube-dl -u BelSean21@somewhere.xxx.au -p password https://www.animelab.com/player/gleipnir-episode-1
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] gleipnir-episode-1: Downloading requested URL
WARNING: [AnimeLab] Could not get height of video
WARNING: [AnimeLab] Could not get height of video
[download] Destination: Gleipnir - Episode 1 - Something Inside of Me-55226.mp4
[download]   1.2% of 455.50MiB at 312.45KiB/s ETA 24:34
```
Doesn't seem to effect the resultant download. Video plays fine

**Comment by inks007** [comment]
@mariuszskon --download-archive looks to be working correctly, thank you.

The warnings seem to depend on the video (probably happens with the newer ones).
```
~/youtube-dl $ python3 -m youtube_dl --netrc https://www.animelab.com/player/arte-episode-3
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] arte-episode-3: Downloading requested URL
WARNING: [AnimeLab] Could not get height of video
WARNING: [AnimeLab] Could not get height of video
[download] Destination: Arte - Episode 3 - First Job-55155.mp4
```
I tried my-hero-academia-episode-1 like you did (it is a very old episode) and there aren't any warnings there.

**Comment by mariuszskon** [comment]
Hmm this is a very interesting problem. @BelSean21 Gleipnir is working correctly on my end. @inks007 I cannot test your link because it is too new and requires a premium account, which I do not have. Episode 2 seems to work fine.
Since the warning is appearing twice, that means two of the formats do not have an exposed height but are otherwise considered valid by the extractor (which afaik means they have a URL). If you use `--list-formats` you can usually see more than two formats.
After writing most of this I realised what the problem is - since I do not have premium, I do not have access to the same formats you do. This is good because we can deal with this edge case right here :smile: 
To minimise the amount of data for me to sift through, and to avoid leaking URLs with copyright information, please add the following line after line 156 in `youtube_dl/extractors/animelab.py`:
```
print(quality_data)
```
(make sure it has the right indentation i.e. it has the same number of spaces as line before it.

Then you can post the results of attempting to get those height-less videos :smile:

**Comment by inks007** [comment]
@mariuszskon Sent you a subscription code instead. Much easier.

**Comment by BelSean21** [comment]
@mariuszskon  Like this?

```
pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $ ./youtube-dl -u BelSean21@somewhere.xxx.au -p password  https://www.animelab.com/player/gleipnir-episode-1
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] gleipnir-episode-1: Downloading requested URL
{u'videoQualityType': u'SD', u'description': u'360p', u'videoFormat': {u'videoFormatType': u'PROGRESSIVE', u'name': u'MP4'}, u'id': 1, u'name': u'360p'}
{u'videoQualityType': u'HD', u'description': u'720p', u'videoFormat': {u'videoFormatType': u'PROGRESSIVE', u'name': u'MP4'}, u'id': 3, u'name': u'720p'}
{u'videoQualityType': u'SD', u'description': u'480p', u'videoFormat': {u'videoFormatType': u'PROGRESSIVE', u'name': u'MP4'}, u'id': 2, u'name': u'480p'}
{u'videoQualityType': u'HD', u'description': u'1080p', u'videoFormat': {u'videoFormatType': u'PROGRESSIVE', u'name': u'MP4'}, u'id': 4, u'name': u'1080p'}
{u'videoQualityType': u'SD', u'description': u'SD HLS', u'videoFormat': {u'videoFormatType': u'ADAPTIVE', u'name': u'HLS'}, u'id': 6, u'name': u'SD HLS'}
{u'videoQualityType': u'HD', u'description': u'HD HLS', u'videoFormat': {u'videoFormatType': u'ADAPTIVE', u'name': u'HLS'}, u'id': 7, u'name': u'HD HLS'}
WARNING: [AnimeLab] Could not get height of video
{u'videoQualityType': u'SD', u'description': u'SD MPEG-DASH', u'videoFormat': {u'videoFormatType': u'ADAPTIVE', u'name': u'MPEG-DASH'}, u'id': 8, u'name': u'SDDASH'}
{u'videoQualityType': u'HD', u'description': u'HD MPEG-DASH', u'videoFormat': {u'videoFormatType': u'ADAPTIVE', u'name': u'MPEG-DASH'}, u'id': 9, u'name': u'HDDASH'}
WARNING: [AnimeLab] Could not get height of video
[download] Resuming download at byte 12167225
[download] Destination: Gleipnir - Episode 1 - Something Inside of Me-55226.mp4
[download]   3.3% of 455.50MiB at 897.90KiB/s ETA 08:22?
ERROR: Interrupted by user
pi@raspberrypi:~/Projects/youtube-dl_Animelab/youtube-dl $
```

**Comment by BelSean21** [comment]
@mariuszskon Getting this with some of the Movies 

```
pi@raspberrypi:~/Videos $ ./youtube-dl -u BelSean21@somewhere.xxx.au -p password https://www.animelab.com/player/hentai-kamen
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] hentai-kamen: Downloading requested URL
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "./youtube-dl/__main__.py", line 19, in <module>
  File "./youtube-dl/youtube_dl/__init__.py", line 474, in main
  File "./youtube-dl/youtube_dl/__init__.py", line 464, in _real_main
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 2018, in download
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 796, in extract_info
  File "./youtube-dl/youtube_dl/extractor/common.py", line 530, in extract
  File "./youtube-dl/youtube_dl/extractor/animelab.py", line 135, in _real_extract
AttributeError: 'NoneType' object has no attribute 'get'

pi@raspberrypi:~/Videos $ ./youtube-dl -u BelSean21@somewhere.xxx.au -p password https://www.animelab.com/player/attack-on-titan-live-action-part-1
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] attack-on-titan-live-action-part-1: Downloading requested URL
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "./youtube-dl/__main__.py", line 19, in <module>
  File "./youtube-dl/youtube_dl/__init__.py", line 474, in main
  File "./youtube-dl/youtube_dl/__init__.py", line 464, in _real_main
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 2018, in download
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 796, in extract_info
  File "./youtube-dl/youtube_dl/extractor/common.py", line 530, in extract
  File "./youtube-dl/youtube_dl/extractor/animelab.py", line 135, in _real_extract
AttributeError: 'NoneType' object has no attribute 'get'
pi@raspberrypi:~/Videos $
```

**Comment by mariuszskon** [comment]
@inks007 you are extremely generous! I will do my best to take advantage of your donation to make AnimeLab extraction as good as it can be!

@BelSean21 thanks for following my instructions, the output is exactly what I wanted to see. Of course I can myself confirm this is the case thanks to the donation of premium. Basically, if you look carefully, AnimeLab is sometimes reporting the quality as 'HD' instead of '1080p'. This is a simple fix, which I could have hacked together, or, which I realised youtube-dl has more built-in features for analysis :) This has been implemented in c3dca17.

I also noticed that the English or Japanese option is not handled smoothly (it should get both options in a single command), so I fixed that to an extent in 2f6d029. It's not the most elegant solution, but it works.

@BelSean21 I have fixed the movie extraction issue in 2cf8283.

**Comment by inks007** [comment]
@mariuszskon Thanks for all the fixes.
I see that --list-formats now shows all the DASH formats

There is some minor weirdness with picking up different formats:
```
python3 -m youtube_dl --netrc https://www.animelab.com/player/kakushigoto-episode-4 --list-formats
[AnimeLab] Downloading login page
[AnimeLab] Logging in
[AnimeLab] kakushigoto-episode-4: Downloading URL https://www.animelab.com/player/kakushigoto-episode-4/subtitles
[AnimeLab] 55326: Downloading m3u8 information
[AnimeLab] 55326: Downloading m3u8 information
[AnimeLab] 55326: Downloading MPD manifest
[AnimeLab] 55326: Downloading MPD manifest
[AnimeLab] kakushigoto-episode-4: Downloading URL https://www.animelab.com/player/kakushigoto-episode-4/dubbed
[AnimeLab] 55326: Downloading m3u8 information
[AnimeLab] 55326: Downloading m3u8 information
[AnimeLab] 55326: Downloading MPD manifest
[AnimeLab] 55326: Downloading MPD manifest
[info] Available formats for 55326:
format code                               extension  resolution note
157821_yeshardsubbed_ja-JP-audio-ENGLISH  m3u8       audio only [en]
157822_yeshardsubbed_ja-JP-audio-ENGLISH  m3u8       audio only [en]
157823_yeshardsubbed_ja-JP-0              m4a        audio only DASH audio  209k , m4a_dash container, mp4a.40.2 (48000Hz)
157824_yeshardsubbed_ja-JP-0              m4a        audio only DASH audio  209k , m4a_dash container, mp4a.40.2 (48000Hz)
157821_yeshardsubbed_ja-JP-583            m3u8       426x240     583k , avc1.42c01e, video only
157822_yeshardsubbed_ja-JP-583            m3u8       426x240     583k , avc1.42c01e, video only
157823_yeshardsubbed_ja-JP-1              mp4        426x240    DASH video  714k , mp4_dash container, avc1.42c01e, video only
157824_yeshardsubbed_ja-JP-1              mp4        426x240    DASH video  714k , mp4_dash container, avc1.42c01e, video only
157821_yeshardsubbed_ja-JP-725            m3u8       640x360     725k , avc1.640828, video only
157822_yeshardsubbed_ja-JP-725            m3u8       640x360     725k , avc1.640828, video only
157821_yeshardsubbed_ja-JP-911            m3u8       853x480     911k , avc1.640828, video only
157822_yeshardsubbed_ja-JP-911            m3u8       853x480     911k , avc1.640828, video only
157823_yeshardsubbed_ja-JP-2              mp4        640x360    DASH video 1212k , mp4_dash container, avc1.640828, video only
157824_yeshardsubbed_ja-JP-2              mp4        640x360    DASH video 1212k , mp4_dash container, avc1.640828, video only
157822_yeshardsubbed_ja-JP-1264           m3u8       1280x720   1264k , avc1.640828, video only
157823_yeshardsubbed_ja-JP-3              mp4        854x480    DASH video 1997k , mp4_dash container, avc1.640828, video only
157824_yeshardsubbed_ja-JP-3              mp4        854x480    DASH video 1997k , mp4_dash container, avc1.640828, video only
157822_yeshardsubbed_ja-JP-2078           m3u8       1920x1080  2078k , avc1.640829, video only
157824_yeshardsubbed_ja-JP-4              mp4        1280x720   DASH video 3503k , mp4_dash container, avc1.640828, video only
157824_yeshardsubbed_ja-JP-5              mp4        1920x1080  DASH video 4769k , mp4_dash container, avc1.640829, video only
157819_yeshardsubbed_ja-JP                mp4        360p       [ja-JP]
157817_yeshardsubbed_ja-JP                mp4        480p       [ja-JP]
157818_yeshardsubbed_ja-JP                mp4        720p       [ja-JP]
157820_yeshardsubbed_ja-JP                mp4        1080p      [ja-JP]  (best)
```
-f 157824_yeshardsubbed_ja-JP-4 works
-f "mp4[format_id*=ja-JP-4]" works
-f "[format_id*=ja-JP-4]" doesn't work
-f "[format_id*=157818]" works

Can't seem to specify picking up the newer DASH formats without specifying the extension (mp4 or m4a). This is probably minor, because all of the following work:
-f "bestvideo+bestaudio" picks up the best DASH video
-f "bestvideo[height<=720]+bestaudio" picks up the best DASH 720p video
-f "best" picks up the best muxed video

**Comment by mariuszskon** [comment]
@inks007 Indeed, we now have lots more formats thanks to actually correctly handling mpd and m3u8 :sweat_smile: 

The format shown is what my extractor reports to youtube-dl. All control is then passed to the rest of youtube-dl where selecting the format, actually performing the download of the media etc. is actually done. Therefore it is most likely that the problem is not in my extractor. However, I have been wrong before, so I will take a look. My response will be a bit delayed.

**Comment by BelSean21** [comment]
Thankyou @mariuszskon Looking good :-)

**Comment by mariuszskon** [comment]
@inks007 after a bunch of research and experimenting I have come to a few conclusions regarding the issue of format selection.
First, it has nothing to do with the particular extractor (other than the list of formats) but a lot to do with youtube-dl's implementation of format selection. There is a TL;DR at the bottom if you do not care about the technical details.
In `YoutubeDL.py`:
```
        FormatSelector = collections.namedtuple('FormatSelector', ['type', 'selector', 'filters'])
```
Debugging this with your given format specifiers gives the following results:
* `-f 157824_yeshardsubbed_ja-JP-4` => `[FormatSelector(type='SINGLE', selector='157824_yeshardsubbed_ja-JP-4', filters=[])]`
* `-f "mp4[format_id*=ja-JP-4]"` => `[FormatSelector(type='SINGLE', selector='mp4', filters=['format_id*=ja-JP-4'])]`
* `-f "[format_id*=ja-JP-4]"` => `[FormatSelector(type='SINGLE', selector='best', filters=['format_id*=ja-JP-4'])]`
* `-f "[format_id*=157818]"` => `[FormatSelector(type='SINGLE', selector='best', filters=['format_id*=157818'])]`

See the problem? `best` appears seemingly out of nowhere. This is because some `selector` is necessary. The problem with `best` is that it requires audio *and* video to be within the one file. `157824_yeshardsubbed_ja-JP-4` is video only. Therefore `-f "[format_id*=ja-JP-4]"` is equivalent to `-f "best[format_id*=ja-JP-4]"` which understandably fails because there is no combined audio+video file which matches that filter.

Similarly, you will see that `-f "[format_id*=157824]"`, which is in the format of your "working" last example, produces `[FormatSelector(type='SINGLE', selector='best', filters=['format_id*=157824'])]` and does **not** work for `157824_yeshardsubbed_ja-JP-4`. `-f "[format_id*=157818]"` works because the target file does have both audio and video.

**TL;DR:** If you only specify things in brackets `[ ]` then youtube-dl not only wants it to match that filter, but also be the `best`. Sometimes it cannot find a `best` that is also filtered down, because it requires both audio and video in the same file. Solution/workaround: specify *something* as the selector, such as the extension. Or maybe, you do want the particular video and you wish to mux it with the best audio: `-f "bestvideo[format_id*=ja-JP-4]+bestaudio"`

**Comment by mariuszskon** [comment]
@dstftw I understand that you are very busy since youtube-dl is a massive project, but I would love it if you could re-review my code.

**Comment by inks007** [comment]
@mariuszskon that makes a lot of sense, thanks.

**Comment by aristaeus** [comment]
Animelab's being shut down so you may as well close this PR :'(

**Comment by mariuszskon** [comment]
> Animelab's being shut down so you may as well close this PR :'(

Thanks for the update, indeed it is. Apparently you can login to funimation with the same details, see https://www.animelab.com/blog/animelab-is-becoming-funimation-in-australia-and-new-zealand/

Whether you like the move or not, the good part from youtube-dl's perspective is there is less code to maintain :smile:
